eclipse-theia/theia
# refactor core messaging and RPC (electron-only so far)

- Add new reusable components to do communication and RPC in the framework.
- Refactor `electron-main` to `electron-browser` to use those new RPC components.
- Refactor the sample updater from `dev-packages/sample-api` to use the new system.
- Add `electron-main` to `backend` IPC communication.

I open this PR as a draft because **I still need to add much needed documentation and tests**.

I expect from this PR to get feedback on the proposed new APIs and go forward with it. I also hope for https://github.com/eclipse-theia/theia/pull/11011 to align with it.

This PR only covers Electron-specific communication, I left out the `backend` to `frontend` part for now. We can discuss if it should be done in this PR too or not.

Here's a temporary rundown of the proposed new messaging and RPC system, until I document it properly:

---

The goal of the new APIs is to simplify usage by extenders, internal wiring should hold all the complexity but stay reasonable.

`Connection<T>` are the heart of any communication. `T` can be an object, serialization may happen eventually. Connections are often API adapters: Each and every communication library has its own API to send and receive messages but for most relevant use cases only `sendMessage` and `onMessage` are required. You are only required to implement this abstraction if you wish to feed your own details into the other reusable messaging components.

`RpcConnection` are the highest-level abstraction to be able to setup JavaScript proxies. Note that instances _are scoped_ to their respective remote instances, hence why there is no notion of `id` in this API. However you establish a remote connection to a specific instance, or whatever protocol you actually use, the `RpcConnection` is abstracting it.

`ProxyProvider` are named bindings to allow binding a proxy as a regular Inversify service. Example:

```ts
bind(SomeRemoteService)
    .toDynamicValue(ctx => ctx.container.getNamed(ProxyProvider, 'backendOrFrontendOr...').getProxy(remoteServicePath))
    .inSingletonScope();
```

The API is meant to hide how proxies are setup: connection, protocol, etc... You just want a proxy to some remote runtime.

Proxies no longer have a `setClient` method: if you want to react to notifications you can add a listener on methods named like `onSomething...`. If you want a remote object to make requests and get responses, you need a new proxy.

`ServiceContribution` is the new API to expose remote services for proxies to be created. They are either a map of `servicePath` to a callback to get the actual instance to serve over RPC, or directly a callback that will handle the request (`serviceId, serviceParams`).

Inversify bindings might be more verbose now, but this is the best trade-off I was able to find to have small components that make sense. Implementation details are kept to the binding of internals. Theia extenders will most likely never have to write as much wiring code just to use what I documented so far.

Lots of implementation details.

Noteworthy:

- I kept using JSON-RPC as protocol for now, my main concern is redesigning our APIs/abstractions to make JSON-RPC -or any other protocol- a replaceable implementation detail.
- The current JSON-RPC implementation relies on a generic connection multiplexer to serve different remote services over the main connection used for messaging.
- I don't explicitly serialize anything and instead I rely on the low-level transport's implicit serialization. We can serialize ourselves at any point, should we require it.
- Implementation of a simple and generic `ConnectionMultiplexer` that can run over most `Connection<any>`. Supports bi-directional channel opening.
- Inversify only resolves bindings as needed/required. You might see some places where I `get` a binding without doing anything with it, this is because there are side-effects when resolving it (e.g. serve RPC calls).
- We often have an issue with Inversify where some bindings _have_ to be resolved synchronously, but they are inherently constructed asynchronously. This is especially true with some proxy connections that can only be setup once something happens. This is why I rely on `DeferredConnection` in various places: It will buffer messages until the underlying connection is ready, but I can start using it right-away to get a RPC proxy out of it synchronously.
- I added ["tagged strings and symbols"](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#type-aliases) to allow TypeScript to understand our typings when binding using Inversify. See `packages/core/src/common/types.ts#serviceIdentifier`.
- Reacting to the frontend loading and reloading itself in Electron `BrowserWindow`s is a pain...

#### How to test

Run the Electron app, IPC-related features should work:

- You should see a log in the console output with the title of all opened windows.
- You should be able to execute the `Open New Window` command.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
